torch_xla/experimental/megablox/gmm.py

```diff
@@ -0,0 +1,357 @@
+"""Grouped matrix multiplication kernels for TPU written in Pallas."""
+
+from typing import Any, Callable, Optional, Union
+from torch_xla.experimental.megablox import common
+import torch
+import torch_xla
+import numpy as np
+
+
+def jax_import_guard():
+  # Somehow, we need to grab the TPU before JAX locks it. Otherwise, any PT/XLA TPU operations will hang.
+  torch_xla._XLAC._init_computation_client()
+
+
+jax_import_guard()
+import jax
+import jax.numpy as jnp
+
+
+def _validate_args(
+    *,
+    lhs: torch.Tensor,
+    rhs: torch.Tensor,
+    group_sizes: torch.Tensor,
+    expected_rhs_dims: int = 3,
+) -> tuple[jnp.ndarray, jnp.ndarray, jnp.dtype]:
+  """Validates the arguments for the gmm function."""
+  # Validate 'lhs'.
+  if lhs.dim() != 2:
+    raise ValueError(f"Expected 2-tensor for 'lhs' but got {lhs.dim()}-tensor.")
+  common.assert_is_supported_dtype(lhs.dtype)
+
+  # Validate 'rhs'.
+  if rhs.dim() != expected_rhs_dims:
+    raise ValueError(f"Expected {expected_rhs_dims}-tensor for 'rhs' but got"
+                     f" {rhs.dim()}-tensor.")
+  common.assert_is_supported_dtype(rhs.dtype)
+
+  # Validate 'group_sizes'.
+  if group_sizes.dtype != torch.int32:
+    raise ValueError(
+        f"Expected 32-bit integer 'group_sizes' but got {group_sizes.dtype}.")
+
+  return lhs, group_sizes, common.select_input_dtype(lhs, rhs)
+
+
+def _calculate_num_tiles(x: int, tx: int) -> int:
+  tiles, rem = divmod(x, tx)
+  if rem:
+    raise ValueError(f"{x} must be divisible by x-dimension tile size ({tx}).")
+  return tiles
+
+
+def _calculate_irregular_num_tiles(x: int, tx: int) -> tuple[int, int]:
+  tiles, rem = divmod(x, tx)
+  if rem:
+    tiles += 1
+  return tiles, rem
+
+
+GroupMetadata = Any  # TODO(enriqueps): Clean this up and use a namedtuple
+
+
+def _make_group_metadata(
```

Comment:
Why can't we just use the JAX version?

Reply:
When we use the jax version to do this compute. Let's make sure we pass jax CPU tensors such that this part of compute can be done in cpu instead. As far as I can tell only group_sizes is used and it's 1D, so should be pretty lightweight to compute. We should also benchmark this against the reference_gmm in case this part drastically increase the tracing time. On that matter, we should cache the result.

Reply:
Actually we can't do this given the group_sizes is data produced in the middle of the graph. And it means we need to do a graph break.